### PR TITLE
feat(interaction): implement multi-touch click suppression for ButtonInteractionHandler

### LINKED ISSUE #579: feat: add multitouch support to MouseEventManager and ButtonInteractionHandler
# Add Multitouch Support to MouseEventManager and ButtonInteractionHandler

## Problem Statement

The current MouseEventManager implementation uses PointerEvent API but does not utilize the `pointerId` property, making it unable to handle multiple simultaneous touch interactions. This limitation prevents the library from supporting modern multitouch interfaces on tablets and touch-enabled devices.

#### MouseEventManager Constraints
- **Single hover state**: `currentOver: IClickableObject3D<unknown>[]` array cannot distinguish between different pointers (line 121)
- **Unused pointer ID**: PointerEvent `pointerId` property is extracted but never used in event processing
- **Collision detection**: Multiple touches on the same object cannot be differentiated

#### ButtonInteractionHandler Constraints  
- **Boolean state flags**: `_isOver: boolean` and `_isPress: boolean` can only track single pointer state (lines 241, 235)
- **Click detection**: Cannot distinguish same-pointer down→up sequences from cross-pointer events
- **State management**: No mechanism to track which specific pointers are hovering or pressing

#### Existing Infrastructure
- PointerEvent listeners already registered in MouseEventManager (lines 68-70)
- PointerEvent objects received by event handlers (lines 84, 126, 292)
- ViewPortUtil fully supports PointerEvent coordinate transformation

#### MouseEventManager Event Flow
1. Canvas receives PointerEvent
2. `getIntersects()` performs raycasting
3. `checkTarget()` searches for `IClickableObject3D` implementations  
4. `onButtonHandler()` delegates to ButtonInteractionHandler
5. **Issue**: Pointer ID lost during delegation process

#### ButtonInteractionHandler State Management
```typescript
// Current single-pointer implementation
protected _isPress: boolean = false;  // Line 235
protected _isOver: boolean = false;   // Line 241

// Compatibility getters
get isOver(): boolean { return this._isOver; }  // Line 185
get isPress(): boolean { return this._isPress; } // Line 196
```

#### PointerEvent API Support
- **Mouse interactions**: Typically `pointerId = 1` (consistent across browsers)
- **Touch interactions**: Sequential ID assignment (1, 2, 3...) but implementation varies
- **Pen/Stylus**: Device-specific ID patterns
- **Investigation required**: Cross-browser pointer ID behavior verification needed

### Dual-Component Architecture Approach

> [!IMPORTANT]
> Both MouseEventManager AND ButtonInteractionHandler require modifications for complete multitouch support.

#### MouseEventManager Responsibilities
- Extract `pointerId` from PointerEvent objects
- Manage per-pointer hover state using `Map<number, IClickableObject3D[]>`
- Delegate events with pointer ID information to handlers
- Handle pointer cleanup when pointers are removed

#### ButtonInteractionHandler Responsibilities  
- Replace boolean flags with pointer ID arrays
- Implement same-ID click detection logic
- Maintain backward compatibility through getter methods
- Manage per-pointer press and hover state

#### ButtonInteractionHandler State Management Overhaul
```typescript
// Proposed multi-pointer implementation
protected hoverPointerIds: number[] = [];
protected pressPointerIds: number[] = [];

// Backward compatibility getters
get isOver(): boolean { return this.hoverPointerIds.length > 0; }
get isPress(): boolean { return this.pressPointerIds.length > 0; }

// Enhanced event handlers with pointer ID support
public onMouseDownHandler(event: ThreeMouseEvent<Value>, pointerId: number = 1): void
public onMouseUpHandler(event: ThreeMouseEvent<Value>, pointerId: number = 1): void  
public onMouseOverHandler(event: ThreeMouseEvent<Value>, pointerId: number = 1): void
public onMouseOutHandler(event: ThreeMouseEvent<Value>, pointerId: number = 1): void
```

#### MouseEventManager Multi-Pointer State Management
```typescript
// Current single-pointer hover tracking
protected currentOver: IClickableObject3D<unknown>[] = [];

// Proposed per-pointer hover tracking  
protected currentOver: Map<number, IClickableObject3D<unknown>[]> = new Map();

// Feature flag for gradual adoption
protected enableMultiTouch: boolean = false; // Default disabled for compatibility
```

#### Click Detection Logic Requirements
- **Same-ID requirement**: Click events only fire when same `pointerId` performs down→up sequence
- **Cross-pointer prevention**: Different pointer IDs performing down→up should NOT trigger click
- **State isolation**: Each pointer's hover/press state managed independently

#### Three Approaches for PointerID Handling

Based on real browser testing (Chrome desktop shows consistent `pointerId = 1` for mouse), three strategic approaches exist for handling undefined or missing pointer IDs:

##### Approach 1: Test Environment Default ID Assignment
**Strategy**: Mock/test environments assign `pointerId = 1` by default when creating synthetic events

**Pros:**
- Minimal production code changes required
- Test infrastructure isolation prevents runtime complexity
- Clean separation between test and production behavior
- Existing test suites require minimal modifications

**Cons:**
- Test-production behavior divergence may hide real issues
- Synthetic event creation becomes more complex
- Potential for missing edge cases in testing
- Mock events may not accurately represent browser behavior

**Implementation:**
```typescript
// In ThreeMouseEventUtil.generate()
export const generate = (type: string, target: any, pointerId: number = 1): ThreeMouseEvent => {
  // Always provide default pointerId in test utilities
  return { type, target, pointerId, /* other properties */ };
}
```

##### Approach 2: MouseEventManager-Level ID Normalization (Recommended)
**Strategy**: MouseEventManager assigns `pointerId = 1` when receiving undefined/missing IDs from events

**Pros:**
- Centralized ID handling with single point of control
- Clean separation of concerns (MouseEventManager handles ID normalization)
- Runtime safety without test-production divergence
- Preserves authentic browser event data until normalization point
- Easy to debug and monitor ID assignment behavior

**Cons:**
- Additional runtime processing overhead (minimal)
- Potential masking of legitimate undefined ID scenarios
- Requires careful logging for debugging purposes

**Implementation:**
```typescript
// In MouseEventManager
protected onDocumentMouseMove = (event: PointerEvent) => {
  // Normalize pointer ID with fallback
  const pointerId = event.pointerId ?? 1;
  
  if (event.pointerId === undefined) {
    console.warn('PointerEvent missing pointerId, using default: 1');
  }
  
  // Continue with normalized ID
  this.processPointerEvent(event, pointerId);
};
```

##### Approach 3: Native Undefined ID Handling
**Strategy**: Preserve undefined pointerID as a distinct state throughout the entire system

**Pros:**
- Most accurate representation of actual browser behavior
- No information loss from original events
- Explicit handling of all possible states
- Potential for advanced debugging and analytics

**Cons:**
- Complex state management throughout entire system
- Higher potential for undefined-related runtime errors
- Requires extensive null checking in all components
- More complex testing scenarios required

**Implementation:**
```typescript
// In ButtonInteractionHandler
protected hoverPointerIds: (number | undefined)[] = [];
protected pressPointerIds: (number | undefined)[] = [];

public onMouseDownHandler(event: ThreeMouseEvent<Value>, pointerId?: number): void {
  // Handle undefined as legitimate pointer ID
  if (!this.pressPointerIds.includes(pointerId)) {
    this.pressPointerIds.push(pointerId);
  }
}
```

#### Recommended Approach: MouseEventManager-Level Normalization

**Rationale:**
- Provides optimal balance between simplicity and accuracy
- Centralizes compatibility logic in appropriate architectural layer
- Maintains clean interfaces for ButtonInteractionHandler
- Enables comprehensive logging and monitoring of ID assignment
- Minimizes risk while preserving debugging capability

**Implementation Strategy:**
```typescript
// MouseEventManager constructor option
constructor(scene, camera, canvas, options?: {
  enableMultiTouch?: boolean;
  fallbackPointerId?: number; // Default: 1
}) {
  this.enableMultiTouch = options?.enableMultiTouch ?? false;
  this.fallbackPointerId = options?.fallbackPointerId ?? 1;
}

// ID extraction with logging
private extractPointerId(event: PointerEvent): number {
  if (event.pointerId !== undefined) {
    return event.pointerId;
  }
  
  // Log for debugging and analytics
  console.warn(`PointerEvent missing pointerId, using fallback: ${this.fallbackPointerId}`, event);
  return this.fallbackPointerId;
}
```

#### API Preservation
- All existing public methods maintain identical signatures
- Default parameter values ensure single-pointer behavior (pointerId = 1)
- Existing getter properties (`isOver`, `isPress`) function unchanged
- Feature flag (`enableMultiTouch: false`) maintains current behavior by default

#### Migration Path
- Existing code requires no changes
- New applications can opt-in with `enableMultiTouch: true`
- Gradual adoption possible through feature flag
- Advanced users can customize `fallbackPointerId` if needed

#### Pointer ID Behavior Verification
- **Mouse consistency**: Verify `pointerId = 1` across Chrome, Firefox, Safari, Edge
- **Touch ID patterns**: Document ID assignment algorithms on mobile browsers
- **Concurrent limits**: Test maximum simultaneous pointer support per browser
- **Edge cases**: Rapid touch/release, pointer persistence across page interactions

### Real Device Investigation Results

Based on comprehensive testing with the custom PointerEvent investigation tool, the following behavior patterns have been confirmed:

#### Desktop Chrome (macOS)
**Mouse Interactions:**
```
Type: mouse
Primary: true (always)
PointerID: 1 (fixed, never changes)
Pressure: 0.000 (movement) / 0.500 (click)
Size: 1x1 (fixed)
Tilt: (0,0) (no tilt data for mouse)
```

**Key Characteristics:**
- **Consistent ID**: `pointerId = 1` for all mouse interactions
- **Reliable Primary**: Always `isPrimary: true`
- **Predictable Pressure**: Binary pressure indication (0.0 or 0.5)
- **Event Sequence**: Direct `POINTERDOWN` → `POINTERMOVE` → `POINTERUP`

#### iPad Safari (iOS)
**Touch Interactions:**
```
Multiple simultaneous touches with distinct characteristics:
Touch 1: pointerId: -560913604, isPrimary: false
Touch 2: pointerId: -560913605, isPrimary: false  
Touch 3: pointerId: -560913606, isPrimary: true
```

**Key Characteristics:**
- **Large Negative IDs**: Unexpected negative values in the hundreds of millions
- **Sequential Assignment**: IDs increment by 1 (-604 → -605 → -606)
- **Non-Primary First Touch**: First touch is NOT primary (differs from expectations)
- **Primary Assignment**: Third touch became primary in test scenario
- **Complex Event Sequence**: `POINTEROVER` → `POINTERENTER` → `POINTERDOWN` (vs. mouse)
- **Pointer Capture Events**: `GOTPOINTERCAPTURE` / `LOSTPOINTERCAPTURE` (touch-specific)

#### Implementation Implications

**ID Range Assumptions (Invalid):**
```typescript
❌ if (pointerId < 1 || pointerId > 100) { /* error */ }
✅ // Accept any number, including large negative values
```

**Primary Pointer Handling (Irrelevant for this project):**
- Primary detection varies by device and interaction sequence
- Not needed for basic button/touch interactions (out of scope)

**Cross-Device Compatibility:**
- Desktop: Reliable small positive integers (typically 1)
- Mobile: Large negative integers with sequential patterns
- Both: Fully compatible with Map<number, T> and array.includes(pointerId)

### PointerID API Specification (MDN Reference)

According to Mozilla Developer Network documentation, `pointerId` values have the following characteristics:

- **Arbitrary Values**: IDs can be any number (positive, negative, large, small)
- **Device-Dependent**: Implementation varies by device and browser
- **No Sequential Guarantee**: IDs may not be sequential (1,2,3...) or predictable
- **Random Assignment Possible**: Some devices assign completely random values
- **Only Guarantee**: Same pointer's event sequence (down→move→up→leave) maintains identical ID

**Source**: https://developer.mozilla.org/en-US/docs/Web/API/PointerEvent/pointerId

#### Implementation Guidelines Based on API Specification

**Safe ID Handling Patterns:**
```typescript
✅ // Use ID as opaque key only
const state = this.pointerStates.get(pointerId);

✅ // Safe array operations
if (this.activeIds.includes(pointerId)) { /* ... */ }

✅ // Map-based state management
this.currentOver.set(pointerId, objects);

✅ // Same-pointer detection (API guaranteed)
const wasPressed = this.pressPointerIds.includes(pointerId);
if (wasPressed) {
  // Reliable same-pointer click detection
}
```

**Prohibited ID Operations:**
```typescript
❌ // Never assume ID ranges or patterns
if (pointerId > 0 && pointerId < 100) { /* ... */ }

❌ // Never assume sequential assignment  
const nextId = pointerId + 1;

❌ // Never use ID values for logic
if (pointerId % 2 === 0) { /* even ID logic */ }

❌ // Never assume ID ordering
if (pointerId1 < pointerId2) { /* first vs second touch */ }
```

**Validation of Real Device Data:**
- Desktop Chrome `pointerId = 1`: Consistent but implementation-specific
- iPad Safari negative IDs (-560913604): Valid per specification, not anomalous
- Sequential increment pattern: Device-specific behavior, not guaranteed across devices

#### Browser Compatibility Matrix Creation
Create comprehensive documentation of:
- Pointer ID reliability per browser/platform combination
- Known inconsistencies or bugs in PointerEvent implementation
- Fallback strategies for unsupported scenarios
- Performance characteristics with multiple concurrent pointers

### Investigation Test Requirements
- Create standalone test page for pointer ID behavior analysis
- Document findings in browser compatibility matrix
- Identify any browser-specific workarounds needed
- Establish reliable cross-browser testing methodology

### Phased Development Approach

This feature will be implemented in three distinct phases, each corresponding to a separate Pull Request to ensure incremental progress and risk management:

#### Phase 1: PointerID Default Value Standardization
**Objective**: Establish consistent pointerId default value (1) across test and production environments

**Scope**:
- Enhance `ThreeMouseEventUtil.generate()` with optional `pointerId` parameter (default: 1)
- Verify backward compatibility with existing test suites
- Prepare test infrastructure for multitouch scenarios
- Update TypeScript type definitions as needed

**Implementation**:
```typescript
// ThreeMouseEventUtil.ts enhancement
export const generate = (
  type: keyof ThreeMouseEventMap, 
  target: any, 
  pointerId: number = 1  // Standardized default value
): ThreeMouseEvent => {
  return { type, target, pointerId, /* other properties */ };
}
```

**Success Criteria**:
- All existing tests pass without modification
- New API supports explicit pointerId specification
- Consistent default behavior across test environments
- Foundation for future fake pointer module independence

**Dependencies**: None
**Deliverable**: Single PR with complete backward compatibility

#### Phase 2: Multitouch Test Scenarios (TDD)
**Objective**: Create comprehensive failing test suite for multitouch functionality using Test-Driven Development

**Scope**:
- `__test__/ButtonInteractionHandler.multitouch.spec.ts` - Array-based state management tests
- `__test__/MouseEventManager.multitouch.spec.ts` - Map-based hover state tests
- `__test__/PointerEvent.browser-behavior.spec.ts` - Real device behavior validation
- Performance tests with multiple concurrent pointers
- Edge case validation (negative IDs, large values, random assignment)

**Test Categories**:
```typescript
// Multi-pointer state management tests
describe('ButtonInteractionHandler Multitouch Support', () => {
  it('should handle multiple pointer IDs independently', () => {
    // Test with realistic iPad-style negative IDs
    const touch1 = ThreeMouseEventUtil.generate('down', mockTarget, -560913604);
    const touch2 = ThreeMouseEventUtil.generate('down', mockTarget, -560913605);
    // Should fail until Phase 3 implementation
  });
  
  it('should detect same-ID click correctly', () => {
    // Same pointerId down→up sequence should trigger click
    // Different pointerId down→up should NOT trigger click
  });
  
  it('should isolate different pointer interactions', () => {
    // Hover/press states should be independent per pointer
  });
});

// Map-based state management tests
describe('MouseEventManager Multitouch Support', () => {
  it('should track per-pointer hover states using Map', () => {
    // currentOver: Map<number, IClickableObject3D[]>
  });
  
  it('should handle pointer cleanup correctly', () => {
    // Verify state cleanup when pointers are removed
  });
});
```

**Success Criteria**:
- All multitouch tests fail as expected (pre-implementation)
- Test scenarios cover identified requirements comprehensively
- Realistic device behavior patterns included (Chrome mouse, iPad touch)
- CI/CD integration for continuous test execution

**Dependencies**: Phase 1 completion
**Deliverable**: Single PR with intentionally failing comprehensive test suite

#### Phase 3: Multitouch Implementation
**Objective**: Implement actual multitouch functionality to satisfy all test requirements

**Scope**:
- Convert `MouseEventManager.currentOver` from array to `Map<number, IClickableObject3D[]>`
- Implement array-based state management in `ButtonInteractionHandler`
- Add pointerId parameter support throughout event processing pipeline
- Maintain complete backward compatibility for existing single-pointer behavior

**Core Implementation**:
```typescript
// MouseEventManager.ts modifications
protected currentOver: Map<number, IClickableObject3D[]> = new Map();

private extractPointerId(event: PointerEvent): number {
  return event.pointerId; // WebGL-capable browsers guarantee this exists
}

// ButtonInteractionHandler.ts modifications  
protected hoverPointerIds: number[] = [];
protected pressPointerIds: number[] = [];

// Backward-compatible getters
get isOver(): boolean { return this.hoverPointerIds.length > 0; }
get isPress(): boolean { return this.pressPointerIds.length > 0; }
```

**Success Criteria**:
- All tests (existing + new multitouch) pass completely
- Existing mouse interactions work identically to current behavior
- Multitouch interactions work as specified in test scenarios
- No performance degradation for single-pointer usage
- Memory usage scales linearly with active pointer count

**Dependencies**: Phase 2 completion
**Deliverable**: Single PR with full multitouch feature implementation

### Phase Dependencies and Risk Management

**Incremental Foundation Building**:
- Phase 1 establishes testing infrastructure without risk
- Phase 2 defines comprehensive requirements through failing tests
- Phase 3 implements against well-defined test criteria

**Rollback Strategy**:
- Each phase can be reverted independently if issues arise
- Clear phase boundaries enable precise problem isolation
- Backward compatibility maintained throughout all phases

**Validation Approach**:
- Continuous integration validates each phase thoroughly
- Real device testing confirms theoretical implementation
- Performance benchmarking ensures no regressions

#### Test Environment PointerID Strategy

Based on real device testing and WebGL browser compatibility analysis, test environments should follow a simplified approach:

**Test Event Generation**:
```typescript
// ThreeMouseEventUtil.generate() enhanced for multitouch testing
export const generate = (
  type: keyof ThreeMouseEventMap, 
  target: any, 
  pointerId: number = 1  // Default ID for test consistency
): ThreeMouseEvent => {
  return { type, target, pointerId, /* other properties */ };
}
```

**Rationale**:
- **WebGL Compatibility**: Browsers supporting WebGL reliably implement PointerEvent with valid `pointerId`
- **Real-world Data**: 
  - Desktop Chrome: Consistent `pointerId = 1` for mouse
  - iPad Safari: Large negative IDs (-560913604, -560913605, -560913606)
  - Production environments never produce undefined `pointerId`
- **Test Simplicity**: Default `pointerId = 1` ensures backward compatibility while allowing explicit ID specification for multitouch scenarios

**Test Examples**:
```typescript
// Single pointer test (uses default pointerId = 1)
it('handles single pointer interaction', () => {
  const event = ThreeMouseEventUtil.generate('down', mockTarget);
  expect(event.pointerId).toBe(1);
});

// Multi-pointer test (explicit IDs for realistic scenarios)
it('handles multiple touch pointers', () => {
  const touch1 = ThreeMouseEventUtil.generate('down', mockTarget, -560913604);
  const touch2 = ThreeMouseEventUtil.generate('down', mockTarget, -560913605);
  expect(touch1.pointerId).toBe(-560913604);
  expect(touch2.pointerId).toBe(-560913605);
});

// Edge case: Large negative IDs from real iPad testing
it('handles negative pointer IDs correctly', () => {
  const event = ThreeMouseEventUtil.generate('down', mockTarget, -999999999);
  expect(handler.processPointer(event)).toBeTruthy();
});
```

#### Core Test Categories

**Multi-Pointer State Management**
- Multiple pointer IDs simultaneously hovering same object
- Independent press/hover state per pointer ID
- Pointer ID array manipulation (add/remove/duplicate prevention)
- State cleanup when pointers removed

**Click Detection Logic**
- Same-ID down→up sequence triggers click event
- Different-ID down→up sequence does NOT trigger click
- Multiple simultaneous clicks from different pointers
- Click event payload contains correct pointer information

**Backward Compatibility**
- Single pointer behavior identical to current implementation
- Default parameter handling (pointerId = 1)
- Existing getter behavior unchanged
- Feature flag disabled maintains current functionality

**Browser Integration**
- Cross-browser pointer ID extraction reliability
- Performance with realistic pointer counts (2-10 concurrent)
- Memory usage scaling with active pointer count
- Event cleanup prevents memory leaks

**Backward Compatibility Validation**
- Verify default `pointerId = 1` behavior in single-pointer scenarios
- Test explicit pointer ID specification for multitouch scenarios
- Validate that existing test suites continue to pass without modification
- Confirm consistent behavior between test and production environments

**Real-world Browser Validation**
- Desktop mouse interactions (Chrome: `pointerId = 1`)
- Mobile touch interactions (Safari: large negative IDs like -560913604)
- Multi-touch simultaneous pointer tracking
- Edge cases with extreme pointer ID values

### Integration Testing Requirements
- End-to-end multitouch workflows
- MouseEventManager → ButtonInteractionHandler coordination
- Real browser environment validation (not just mocked events)
- Touch device testing with actual hardware

#### `src/interactionHandler/ButtonInteractionHandler.ts`
**Lines requiring modification:**
- Line 235: `protected _isPress: boolean` → array-based implementation
- Line 241: `protected _isOver: boolean` → array-based implementation  
- Lines 185, 196: Getter methods updated for array-based compatibility
- Event handler methods: Add optional `pointerId` parameter

**New functionality required:**
- Pointer ID array management methods
- Same-ID click detection logic
- Debug/inspection methods for testing

#### `src/MouseEventManager.ts`
**Lines requiring modification:**
- Line 121: `currentOver` array → Map-based per-pointer tracking
- Lines 254-278: `onDocumentMouseMove` pointer ID extraction
- Lines 284-289: `clearOver` per-pointer clearing capability
- Lines 424-460: `checkTarget` pointer ID parameter support
- `onButtonHandler` static method: pointer ID delegation

**New functionality required:**
- `enableMultiTouch` constructor option
- Pointer ID extraction with fallback logic (recommended: MouseEventManager-level normalization)
- Per-pointer state management and cleanup
- `fallbackPointerId` configuration option for advanced users
- Comprehensive logging for undefined pointer ID scenarios

#### `src/ThreeMouseEventUtil.ts`
**Lines requiring modification:**
- Event generation method: Add optional `pointerId` parameter with default value 1
- Maintain backward compatibility for existing test suites

**Implementation:**
```typescript
export const generate = (
  type: keyof ThreeMouseEventMap, 
  target: any, 
  pointerId: number = 1
): ThreeMouseEvent => {
  return { type, target, pointerId, /* other properties */ };
}
```

#### Demo and Documentation Updates  
- Demo page examples demonstrating multitouch interactions
- API documentation updates for new multitouch capabilities
- JSDoc updates with browser compatibility notes

### Optimization Requirements
- Pointer ID array operations must be efficient (typical use: 2-5 concurrent pointers)
- Map-based hover state should not degrade single-pointer performance
- Memory usage should scale linearly with active pointer count
- Event processing latency should remain under current 33ms throttling threshold

### Performance Testing Criteria
- Benchmark single-pointer performance (must not regress)
- Stress test with 10+ concurrent pointers
- Memory leak verification for pointer cleanup
- Cross-browser performance consistency validation

### Functional Requirements
- Multiple touch pointers can simultaneously interact with different objects
- Same-pointer down→up sequences trigger click events correctly  
- Cross-pointer interactions do not interfere with each other
- All existing mouse functionality preserved without changes

### Technical Requirements
- Complete backward compatibility maintained
- Browser investigation provides reliable compatibility matrix
- Performance acceptable across target browsers and devices
- Comprehensive test coverage for all multitouch scenarios

### Integration Requirements
- MouseEventManager and ButtonInteractionHandler coordinate correctly
- Feature flag allows gradual adoption in existing applications
- Demo pages demonstrate practical multitouch usage patterns
- Documentation updated with implementation and browser notes

### Key Code Locations
- **MouseEventManager.ts:121** - `currentOver` array declaration
- **MouseEventManager.ts:68-70** - PointerEvent listener registration  
- **ButtonInteractionHandler.ts:235,241** - Boolean state flags
- **ButtonInteractionHandler.ts:185,196** - State getter methods
- **MouseEventManager.ts:254-278** - `onDocumentMouseMove` implementation
- **MouseEventManager.ts:424-460** - `checkTarget` method

### Related Test Files
- **`__test__/ButtonInteractionHandler.spec.ts`** - Existing handler tests for reference
- **`__test__/MouseEventManager.spec.ts`** - Existing manager tests for compatibility
- **`__test__/MouseEventManager.overlapping.spec.ts`** - Complex interaction scenarios

### Browser API References
- [PointerEvent API Specification](https://developer.mozilla.org/en-US/docs/Web/API/PointerEvent)
- [Pointer Events Browser Compatibility](https://caniuse.com/pointer)
- [Touch Events vs Pointer Events](https://developer.mozilla.org/en-US/docs/Web/API/Pointer_events#comparing_to_touch_events)

### PR DESCRIPTION
**Pull request by MasatoMakino**
## Summary
Implements multi-touch click suppression to prevent unintended click events when multiple pointers interact with the same object simultaneously.

## Motivation
Closes #579 - Multi-touch interactions on ButtonInteractionHandler cause unintended multiple click events

Multi-touch devices allow users to accidentally trigger the same interactive object with multiple fingers, leading to duplicate click events and poor user experience. This implementation ensures only the first pointer release triggers a click event, suppressing subsequent releases.

## Out of Scope
- **Touch gesture recognition** - Swipe, pinch, and other complex gestures not implemented
- **Performance optimization** - High-frequency pointer event optimization deferred  
- **Visual feedback** - Multi-touch visual states not included in this implementation
- **Other interaction handlers** - CheckBox and RadioButton handlers unchanged, maintain existing behavior

### Target
- ButtonInteractionHandler multi-touch state management with `pressPointerIds: Set<number>`
- Click suppression logic for subsequent pointer releases after first click
- Complete press state cleanup after first click in multi-touch scenarios
- Test suite expansion with comprehensive multi-touch interaction coverage

### Dependencies  
- ThreeMouseEventUtil API migration to `createThreeMouseEvent()` function
- Existing mouse event processing pipeline (extended, not replaced)
- Current single-touch behavior preservation for backward compatibility

### Boundaries
- Changes limited to ButtonInteractionHandler click logic only
- No modifications to hover/press visual state management
- Mouse event delivery system unchanged (MouseEventManager unaffected)
- Pointer tracking isolated to individual handler instances

### Focus Areas
- Multi-touch click suppression logic correctness
- Single-touch backward compatibility preservation  
- Memory management for pointer tracking collections
- Test coverage for edge cases and mixed interaction scenarios

## Summary by CodeRabbit

- New Features
  - Improved multi-touch support across buttons, checkboxes and radios: per-finger hover/press tracking, robust click suppression, and synchronized selection state for simultaneous pointers.

- Refactor
  - Event creation and interaction-scanning surfaced via a simpler public API; the interaction-scanning flag was renamed but remains backward-compatible; behavior preserved.

- Style
  - Demos: improved touch UX by disabling long-press text selection, tap highlight, and touch callout on canvases.

### CODE REVIEW

File: src/interactionHandler/ButtonInteractionHandler.ts [hunk 1]
```diff
@@ -377,49 +434,65 @@ export class ButtonInteractionHandler<Value> extends EventEmitter<
...
 
-      const e = ThreeMouseEventUtil.generate("click", this, event.pointerId);
+      const e = createThreeMouseEvent("click", this, event.pointerId);
       this.emit(e.type, e);
+
+      // Multi-touch click suppression: Clear all remaining press states
+      // to prevent subsequent pointers from triggering click events
+      this.pressPointerIds.clear();
     }
   }
```
[the remark this answers is not in the document]
Comment (MasatoMakino) [reply src/interactionHandler/ButtonInteractionHandler.ts:497]: this.isOverは単なるフラグではなく、pressPointerIdsを参照するgetter関数です。

src/interactionHandler/ButtonInteractionHandler.ts:196~198

一つでもpointerが残っていればtrueを返します。したがって、他のアクティブなポインタを無視するという懸念には当たりません。

File: src/interactionHandler/ButtonInteractionHandler.ts [hunk 2]
```diff
@@ -472,13 +545,14 @@ export class ButtonInteractionHandler<Value> extends EventEmitter<
    *
    * @description
    * Manages hover state tracking and visual state transitions for both over and out events.
+   * Supports multiple simultaneous pointers by managing a Set of hover pointer IDs.
    * Hover state tracking occurs unconditionally (even when disabled/frozen) to ensure
    * proper visual updates when the object transitions back to an active state while
-   * the pointer is still hovering over it.
+   * pointers are still hovering over it.
    *
```
[the remark this answers is not in the document]
Comment (MasatoMakino) [reply src/interactionHandler/ButtonInteractionHandler.ts:552]: this.isOverは単なるフラグではなく、pressPointerIdsを参照するgetter関数です。

src/interactionHandler/ButtonInteractionHandler.ts:196~198

一つでもpointerが残っていればtrueを返します。したがって、他のアクティブなポインタを無視するという懸念には当たりません。

File: src/interactionHandler/ButtonInteractionHandler.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (MasatoMakino) [reply src/interactionHandler/ButtonInteractionHandler.ts:497]: calculateCurrentState()を採用するのが正しい修正方針だと考えます。追って修正をコミットします。

File: src/interactionHandler/ButtonInteractionHandler.ts [hunk 3]
```diff
@@ -247,15 +292,27 @@ export class ButtonInteractionHandler<Value> extends EventEmitter<
...
+  /**
+   * @deprecated Use interactionScannable instead. Will be removed in next major version.
+   */
+  public get mouseEnabled(): boolean {
+    return this.interactionScannable;
+  }
+
+  public set mouseEnabled(value: boolean) {
+    this.interactionScannable = value;
+  }
```
[the remark this answers is not in the document]
Comment (MasatoMakino) [reply src/interactionHandler/ButtonInteractionHandler.ts]: mouseEnabledプロパティは将来廃止予定であること、mouseEnabledプロパティに関連するテストシナリオに将来廃止予定であることを明示しました。

**Comment by MasatoMakino** [comment]
Thank you for the question about pointerId normalization and enableMultiTouch option.

  PointerID Fallback Implementation ✅

  The pointerId fallback to 1 is implemented throughout the codebase:

  ThreeMouseEventUtil.ts:
  // Line 120, 179
  pointerId: number = 1,

  MouseEventManager.ts:
  // Lines 547, 601, 749
  pointerId: number = 1,

  Usage Pattern:
  // Real pointer events: uses actual pointerId
  const pointerId = event.pointerId;  // Line 325, 423

  // Fallback scenarios: defaults to 1
  this.checkIntersects(intersects, eventType, pointerId);

  EnableMultiTouch Option: Intentionally Omitted ✅

  This library uses an always-on multitouch design based on a key technical assumption:

  WebGL → PointerEvent Support Correlation:
  - Prerequisite: This library requires WebGL support for Three.js rendering
  - Reality: All modern browsers supporting WebGL also support PointerEvent API
  - Browser Coverage: Chrome 57+, Firefox 59+, Safari 13+, Edge 79+ all support both
  - No Edge Cases: There are no practical scenarios where WebGL works but PointerEvent
  doesn't

  Why no enableMultiTouch option:
  1. Universal Browser Support: PointerEvent API is available in all target browsers
  2. Native Browser Support: Uses standard PointerEvent API
  (pointerdown/pointerup/pointermove)
  3. Zero Configuration: Multitouch works seamlessly without developer intervention
  4. Modern Web Standards: PointerEvent API handles single/multi-touch uniformly
  5. State Management: Built-in per-pointer state tracking via Map<number, 
  IClickableObject3D[]>

  Implementation Evidence:
  // Per-pointer state management (Line 122)
  protected currentOver: Map<number, IClickableObject3D<unknown>[]> = new Map();

  // Automatic pointer extraction from native events
  const pointerId = event.pointerId; // Lines 325, 423

  Design Benefits

  - Seamless Experience: Developers get multitouch support without configuration
  - Backward Compatible: Single-touch interactions work identically (pointerId=1)
  - Future Proof: Leverages native browser multitouch capabilities
  - Simplified API: No additional options to configure or maintain
  - Realistic Assumptions: Based on actual browser capabilities in WebGL environments

  The fallback to pointerId=1 ensures compatibility in test environments and edge cases,
  while the always-on multitouch design provides modern interaction support without
  complexity.
